### PR TITLE
Build System Tweaks for memoized globs

### PR DESCRIPTION
**Pull request by rgrinberg**
To implement memoized globs, I have the following plan:

* Add a glob constructor to `Dep.t` and update all the relevant functions as `dirs`.
* Add a `build_glob` function that would be memoized appropriately.
* Use this function in `build_deps` to build globs.

The main problem that I see that prevents this from working is the current `glob_state` type:

```
  and glob_state =
    | G_unevaluated of Loc.t * Path.t * (Path.t -> bool)
    | G_evaluated   of Path.Set.t
```

We would like to get rid of the above type as we want to have a single glob system everywhere (user deps, library artifacts, external dirs), but the above glob definition is a function that cannot be used as input to a memoized function.

The above predicate definition is too flexible anyway, so I suggest the following:
* We switch the definition to `Predicate_lang.t`
* We change the definition of `Glob.t` to save the `Re.t` along with `Re.re`. This is needed to compare `Glob.t`. Now unfortunately, it's still not enough to compare `Predicate_lang.t`, but it's probably not a big deal.

### CODE REVIEW

**Comment by rgrinberg** [comment]
Note that this will also require me to make some changes in the same spirit to `Predicate_lang.t` as well. None of this is very hard, but I would like to get some agreement before I do all this work.

**Comment by diml** [comment]
Why don't we simply use the string representation of the glob as provided by the user?

File: src/build_system.ml [hunk 1]
```diff
@@ -1219,12 +1221,12 @@ let evaluate_rule (rule : Internal_rule.t) =
 
 (* Same as the function just bellow, but with less opportunity for
    parallelism. We keep this dead code here for documentation purposes
-   as it is eaiser to read the one bellow. The reader only has to
+   as it is easer to read the one bellow. The reader only has to
```
Comment (aalekseyev) [inline src/build_system.ml]: The typo here was not quite fixed. :-)

**Comment by aalekseyev** [comment]
By `Dep.Glob`, do you mean just the listing or the actual files as well?
What about `build_glob`? 
Will it build just the listing (so effectively a filtered and augmented `readdir`) or does it build the individual files matching the pattern too?

**Comment by rgrinberg** [comment]
> Why don't we simply use the string representation of the glob as provided by the user?

There are some globs that are internal. Shall we create a fake string for those?
Anyway, it's probably the same amount of effort. You still need to carry this string representation around everywhere. I'm fine with this as well however.

> By Dep.Glob, do you mean a dependency on just the listing or on the files matching the pattern as well?

Including their contents. I agree it would be useful to differentiate between the two, but for now the user can only express one kind of dependency.

> What about build_glob?

It would include the pattern.

**Comment by rgrinberg** [comment]
Apart from the discussion? Are the changes here ok?

**Comment by aalekseyev** [comment]
Looks good to me. All I can see is a general cleanup though, pretty much unrelated to what's discussed.

**Comment by rgrinberg** [comment]
Yup, there's no ticket for the glob feature so I just hijacked this PR.

**Pull request merged**